RavenDB-14634 various fixes

karmeli87 opened the pull request.

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -2585,7 +2585,7 @@ private async Task<(long Index, object Result)> SendToLeaderAsyncInternal(Comman
         {
             //I think it is reasonable to expect timeout twice of error retry
-            var timeoutTask = TimeoutManager.WaitFor(Engine.OperationTimeout, _shutdownNotification.Token);
+            var timeoutTask = TimeoutManager.WaitFor(TimeSpan.FromMinutes(60), _shutdownNotification.Token);
```
Why is that?

karmeli87 replied:
we try to hunt why we got random timeout failures that cause to 4-5 tests simultaneously to fail

ppekrol replied:
Something that @karmeli87 forgot to write here. This PR will not be merged. Just a pure bug hunting code.

@karmeli87 maybe track how much time we wait with stopwatch and if it is > Engine.OperationTimeout do a console writeline with some debug info? Thoughts?

karmeli87 committed:
RavenDB-14634 various fixes

- timeout on HealthCheck
- avoid spamming cluster log with cleanup command
- properly wait for topology modification

karmeli87 committed:
RavenDB-14634 make EnsureNotPassive more async friendly

ayende commented on src/Raven.Client/Documents/Operations/DatabaseHealthCheckOperation.cs:
```diff
@@ -14,11 +15,16 @@ public RavenCommand GetCommand(DocumentConventions conventions, JsonOperationCon
         {
+            public DatabaseHealthCheckCommand()
+            {
+                Timeout = TimeSpan.FromSeconds(15); // maybe even less?
```
Shouldn't we have a single location to define the default timeout?

ppekrol replied:
No, that would affect too many things. All internal'ish commands have this timeout set manually, In theory we could introduce sth like RavenInternalCommand which would set it by default.

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -468,7 +468,7 @@ public static LicenseStatus GetLicenseStatus(License license)
         }
 
-        public void TryActivateLicense(bool throwOnActivationFailure)
+        public async ValueTask TryActivateLicenseAsync(bool throwOnActivationFailure)
```
Why is this an `async ValueTask`?

karmeli87 replied:
In the common case `TryActivateLicenseAsync` just like `EnsureNotPassiveAsync` are expected to return without actually doing any `async` work.

ayende replied:
See here: https://blog.marcgravell.com/2019/08/prefer-valuetask-to-task-always-and.html

In general, if we have to call `await` internally, that isn't saving a lot.
And if we are going to await internally and outside, there is little point in here.
Also, this isn't likely to be in anything that is in the hot path

karmeli87 replied:
I read this and others few times. TBH I'm still confused :-)

As I understand, the bottom line of that post is to use `ValueTask` _always_ (it is even in the title). But to be aware of the API changes.

Yet, as long as `AsyncTaskMethodBuilder` and friends aren't showing up in the profiler `Task` should be fine to use.

ayende replied:
In general, in our internal API, prefer `Task` unless we know that:

* This is called frequently.
* We expect (and will write code) to handle the completion in a synchronous manner.

For external API:

* If they are likely to complete with no async work, use `ValueTask`

ayende commented on src/Raven.Server/Rachis/RachisLogHistory.cs:
```diff
@@ -379,6 +379,18 @@ private static unsafe string ReadGuid(Table.TableValueHolder entryHolder)
             return guid;
         }
 
+        public bool ContainsGuid(ClusterOperationContext context, string guid)
```
Guid is a specific term, but you are passing just ids here, which can be any strings

karmeli87 replied:
Okay, so would something like: `CommandUniqueId` would be a more proper term?

ayende replied:
Yes

ayende commented on src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:
```diff
@@ -322,21 +322,45 @@ private void LogMessage(string message, Exception e = null, string database = nu
         }
 
+
+        private static string GetGuid(Dictionary<string, long> dic)
```
This is not a guid.

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -2359,15 +2359,15 @@ public void AssignNodesToDatabase(ClusterTopology clusterTopology, DatabaseRecor
         }
 
-        public async Task EnsureNotPassiveAsync(string publicServerUrl = null, string nodeTag = "A", bool skipLicenseActivation = false)
+        public async ValueTask EnsureNotPassiveAsync(string publicServerUrl = null, string nodeTag = "A", bool skipLicenseActivation = false)
```
Why?

karmeli87 committed:
RavenDB-14634 address ValueTask comments

karmeli87 committed:
RavenDB-14634 address guid comments
